### ISSUE TITLE
add :Hardcopy command

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#21629]
> now instead use `:TOhtml` and a browser?

I'm in favor of adding some builtin sugar that saves steps here. E.g. we could ship a `:Hardcopy` command that does the following:

1. generates HTML
2. saves the file to reasonable place (~/Downloads or `stdpath('run')` ?)
3. opens the HTML in system web browser

Then you just hit ctrl-p to print, is that good enough? Step (2) above is the annoying part--finding and opening the file.

_Originally posted by @justinmk in https://github.com/neovim/neovim/issues/21472#issuecomment-1369645909_

**Comment by clason** [comment]
Since `:TOhtml` needs to be rewritten anyway to support extmarks, maybe those can be combined (e.g., `:TOhtml print`)?

**Comment by tarvlad** [comment]
Which behavior is expected if the system doesn't have any browser?

**Comment by CarbonHeartDev** [comment]
I can take care of that!

**Comment by justinmk** [comment]
> Since `:TOhtml` needs to be rewritten anyway to support extmarks, maybe those can be combined (e.g., `:TOhtml print`)?

- `:TOhtml` itself is not a great interface, don't want to plant more roots there. I could imagine a future generalized interface (`:export` ?).
- Some sort of main entrypoint like `:Hardcopy` leaves the door open for a "provider" model I guess, too.

> Which behavior is expected if the system doesn't have any browser?

Unspecified for now.

**Comment by CarbonHeartDev** [comment]
> Which behavior is expected if the system doesn't have any browser?

What about using `troff` as a fallback to handle these cases? It's a simple yet popular typesetting system included in many Linux distros!

The only issue I see is with portability since its installation on Windows is a bit messy (you have to do that trough GNUWin32) but having a fallback to handle the absence of browser is a non-issue in Windows and MacOS since the browser is part of the OS and not an external application.

**Comment by clason** [comment]
My main point is that `:TOhtml` needs a rewrite so it would pay off to consider this in tandem -- whichever is the last function standing. A generic `:Export` sounds good to me, too (could hook into `pandoc`).

In any case, it should be a plugin (like `:Man`) and have sensible platform-specific defaults (for the Main Three).

**Comment by CarbonHeartDev** [comment]
> A generic `:Export` sounds good to me, too (could hook into `pandoc`, too, for PDF export on Linux).

Which file format would you use as the output of this new `:Export` for later conversion by pandoc?

**Comment by clason** [comment]
HTML, probably.

**Comment by CarbonHeartDev** [comment]
Hi, I came up with a [first prototype of the plugin](https://github.com/CarbonHeartDev/neovim/blob/37528ed7a6ebbb28dc9cb803a38fd598f526926f/runtime/plugin/hardcopy.vim), I know it lacks license header and docs but at the moment it's just a POC to see if I'm headed to the correct direction. Other things on which I have to work on the next days are sensible file path defaults for the Main Three, as said by @clason, (at the moment I save the html export in a temporary file) and the `troff` fallback which I proposed.

**Comment by justinmk** [comment]
> came up with a [first prototype of the plugin](https://github.com/CarbonHeartDev/neovim/blob/37528ed7a6ebbb28dc9cb803a38fd598f526926f/runtime/plugin/hardcopy.vim), I know it lacks license header and docs but at the moment it's just a POC to see if I'm headed to the correct direction.

looks like the right direction, though we'd want it to be in Lua.

**Comment by hinell** [comment]
What are use cases for `TOhtml` ?

There thousands of JS utils (including simple editors) that can highlight source code in much more simpler fashion.

**Comment by justinmk** [comment]
> A TS-based solution would be ideal here.

sure, but to avoid confusing @CarbonHeartDev  (or anyone else who picks this up), the near-term goal of providing a :hardcopy replacement can use :TOhtml because that's what we currently have. Later, :TOhtml can be replaced by a TS solution.

**Comment by CarbonHeartDev** [comment]
I saw which user @pynappo made some interesting contribution to my plugin, if he's okay with that we can collaborate on this.

**Comment by Nicola-Bini** [comment]
@CarbonHeartDev Can I also collaborate?

**Comment by CarbonHeartDev** [comment]
Sure @Nicola-Bini!

At the moment the browser based printing system is complete (you can test it by [downloading the plugin from my fork](https://github.com/CarbonHeartDev/neovim/blob/ExportBufferCommand/runtime/plugin/hardcopy.lua), putting it in a directory included in your Neovim runtime path, opening Neovim and launching the command `:Hardcopy`) but any idea to make it better it is welcome.

Currently I'm working on fallback systems for headless machines which cannot open files on a browser. The *nix version of this will be based on `lpstat`, `lpr` and some CLI based text formatting tool (I'm undecided between `groff` and simpler alternatives such as `nl` and `pr`). To date I implemented a function to query the system and get a list of available printers. Now I need to display that list in a floating window allowing the user to pick a printer of his choice and to set basic printing options (number of copies, front-back, line numbering enabled/disabled...) before launching the print job (which in concrete will translate to calling `vim.fm.system()` sending the selected buffer content on a CLI pipeline composed by one or more formatting commands and `lpr` in the end). Once the fallback printing is done for *nix OSs I want to do something similar for Windows systems too but since Windows is a little less "CLI-friendly" I have to investigate if I can obtain good enough results.

In the last days I was pretty busy with my job and my studies and I wasn't able to make meaningful progress so for me any contribution is more welcome now than ever.

**Comment by Nicola-Bini** [comment]
Hi @CarbonHeartDev,

Thank you for the summary. 

I can work on the floating window, do you have a clear list of printing options in mind? If not, I can come up with something using the MacOS printing settings as a reference. Perhaps I could start with [these arguments](https://man7.org/linux/man-pages/man1/lpr.1.html).

![Screenshot 2023-01-26 at 8 38 51 PM](https://user-images.githubusercontent.com/63654846/214991839-4ebfa188-3196-4f1d-a5a7-8adc45af85e7.png)


On another note, I do not have experience, please do not hesitate to point any incorrectness

**Comment by CarbonHeartDev** [comment]
@Nicola-Bini I think which selecting the printer, the number of copies, toggling the line numbering and eventually orientation would be enough, maybe tonight or tomorrow I can write a printing function with said parameters in order to allow you to focus on the GUI while I take care of interfacing with printing and formatting tools.

Paper size selection and double sided print toggle would be nice to have but after hours of research I wasn't able to query CUPS to find which types of paper are supported by a certain printer of if double sided is supported in order to show these on the interface only when actually supported by the selected printer so eventually it's going to be user's responsibility to selecting options which works for his printer and I'm not sure which that's a good UX but I'm open to feedback on this regard and on anything else.

**Comment by Nicola-Bini** [comment]
Hi @CarbonHeartDev,

I made a first draft of the window. I would like some feedback from you. do you have some time to meet on discord/zoom to go over it?

**Comment by leiserfg** [comment]
We need an API to get the precalculated appearance of each character in the buffer, at least background and foreground colors. With that in place we can join contiguous characters with the same "appearance". Starting from that implementing backends for rendering several formats (html, rtf,groff) will be relatively easy  as  we can borrow the "know-how" from pygments. We could even render to PDF for printing in pure lua  by using <https://github.com/catseye/pdf.lua>

**Comment by 42triangles** [comment]
It should be noted that least with the existing `:TOhtml` line numbers & line wrapping do not work together - the wrapped lines start where the number starts. This should likely be somehow addressed for a `:hardcopy` replacement based on `:TOhtml`.

**Comment by hinell** [comment]
The `:Hardcopy` is removed in the latest dev release. I don't think it's worth pursuing this feature.

https://github.com/neovim/neovim/blob/040f1459849ab05b04f6bb1e77b3def16b4c2f2b/runtime/doc/news.txt

**Comment by pynappo** [comment]
> The `:Hardcopy` is removed in the latest dev release. I don't think it's worth pursuing this feature.
> 
> [`040f145`/runtime/doc/news.txt](https://github.com/neovim/neovim/blob/040f1459849ab05b04f6bb1e77b3def16b4c2f2b/runtime/doc/news.txt)

I think it's been removed for a while, and that's why this issue exists? The original comment in this thread has the pull request where it was removed, namely #21472.

**Comment by tjakway** [comment]
I +1 this as an (apparently rare?) user of hardcopy. It was removed without replacement.
I frequently print a bunch of files with :bufdo execute 'hardcopy' and would rather not have to open each HTML in a browser and print it.

In my view any replacement that requires manually clicking anything after :hardcopy is not the same.

I can see how a :TOhtml replacement could work by saving the output to a temp file, running a different command for the html -> pdf step, then calling lpr on it.

But I would rather that existed before hardcopy is scrapped.

**Comment by hinell** [comment]
@tjakway There are standalone tools like [highlight](https://gitlab.com/saalen/highlight) that can be used instead via commandline or inside neovim cmd e.g. `!highlight ... %`

**Comment by 42triangles** [comment]
@tjakway As I'm still trying to search for an alternative that works without breaking with line numbers that I can use as soon as possible, I've also been looking into HTML → PDF conversion tools, and the two that seem to work best are Chromium and its derivates via `--headless --print-to-pdf` (though as far as I can tell you cannot change a lot of options there; also Firefox doesn't seem to have a similar option sadly and if this was going to be the way such a feature would be implemented it would require people to install a Chromium-based browser *solely* for printing) or `weasyprint` (which i should note however is not exactly *fast*, it's okay but calling it for a bunch of files will take a few seconds), since everything else I have tried either doesn't deal with line wrapping at all or has issues with colours.

As a possible solution, until a better one is found, I think the best thing to do is either 1. downgrade the neovim version to one before 0.9.0 for the foreseeable future, 2. switch to `vim` or 3. (the solution I'm probably going to go with) install `vim` alongside `neovim` and call into it for `:hardcopy` purposes.

**Comment by XAUTHORITY** [comment]
Well the new Manjaro update to nvim 0.9 just broke my beloved nvim hardcopy can we keep the hardcopy and related print commands in nvim please?  I don't see the appeal in moving to web browser html printing when I could more quickly print with hardcopy and preview with a 'open in pdf' command with no browser open at all. I honestly use and like it so much that I will be rolling back to 0.8 nvim or going back to vim where these commands still exist.

My setup was as simple as two mappings and two common repo packages. 
```
map <C-p> :hardcopy!<CR>
map <S-p> :hardcopy! > %.ps <Bar> !ps2pdf %.ps && rm %.ps && zathura %.pdf &<CR><CR>
```

Collateral:
```
Notable changes in Nvim 0.9 from 0.8                                     *news*

• `:hardcopy` is now removed (see |hardcopy| and |nvim-removed|):
  - Commands removed:
    - `:hardcopy`
  - Options removed:
    - `printdevice`
    - `printencoding`
    - `printexpr`
    - `printfont`
    - `printheader`
    - `printmbcharset`
```
$ nvim
```
Error detected while processing /home/user/.config/nvim/init.vim:
line  113:
E518: Unknown option: printoptions=paper:A4
line  114:
E518: Unknown option: printoptions=formfeed:n
line  115:
E518: Unknown option: printfont=:h15
line  117:
E518: Unknown option: printoptions=syntax:y
line  118:
E518: Unknown option: printoptions=wrap:y
line  119:
E518: Unknown option: printoptions=duplex:long
line  120:
E518: Unknown option: printoptions=collate:y
line  121:
E518: Unknown option: printoptions=jobsplit:n
line  129:
E518: Unknown option: printoptions=portrait:n,header:0,top:9pc,right:9pc,number:n
```
Thank you for reading, please consider reversing this change.

**Comment by XAUTHORITY** [comment]
right there was also the printoptions that took time to set up how I like it:
```
" PRINT OPTIONS via :hardcopy
set printoptions=paper:A4    "paper size 
set printoptions=formfeed:n  "treat form feed characters (0x0c) as a normal print character 
set printfont=:h15           "font size
"set printoptions=number:n    "display line numbers
set printoptions=syntax:y    "use syntax highlighting
set printoptions=wrap:y      "wrap long lines
set printoptions=duplex:long "print on both sides, bind on long side 
set printoptions=collate:y   "collating 1 2 3, 1 2 3, 1 2 3 
set printoptions=jobsplit:n  "do all copies in one print job 
set wrapmargin=2             "margin settings
"set printoptions=left:0pc    ""goes with portrait line
"set printoptions=right:0pc   ""
"set printoptions=top:10pc     ""
"set printoptions=bottom:0pc  ""
"set printheader=\           "header text 
"set printoptions=portrait:y,header:0  "print orientation portrait, no header
set printoptions=portrait:n,header:0,top:9pc,right:9pc,number:n  "print orientation landscape, no header
```

**Comment by gabrilend** [comment]
So if I understand correctly this workflow was broken without a replacement? I used hardcopy to print my resumes. Printing to HTML doesn't work because it includes a lot of formatting that I didn't need - and line numbers and such. Now I have to go through and fix it manually? That sounds like a lot of work, why was this removed again?

**Comment by Shougo** [comment]
Please see this.

https://github.com/neovim/neovim/pull/15235#issuecomment-1357519408

It is not maintained feature and removed.

**Comment by gabrilend** [comment]
It's not like I *only* need this feature. There are many reasons I don't use Vim.

**Comment by XAUTHORITY** [comment]
unfortunate. at least the 5 hour config migration back to Vim today is mostly where I need it. was always going to be a gamble building good config on a fork. 

:hardcopy Vim is way too convinent to pass up for me.

**Comment by michael-odonovan** [comment]
> 

man this "just go back to vim" stuff is actually supposedly the roots of the neovim project in reverse. Gotta laugh

**Comment by michael-odonovan** [comment]
If the direction of neovim is to scratch features people are clearly using, and then saying "just go back to vim if you don't like it", this project is taking a different turn to what drew me to use neovim...

**Comment by lewis6991** [comment]
> > 
> 
> man this "just go back to vim" stuff is actually supposedly the roots of the neovim project in reverse. Gotta laugh

If you:
- Fix hardcopy for windows
- Make it work for treesitter and extmarks
- Make it easy to maintain

Then we can add it back.

Looking forward to you PR's

**Comment by clason** [comment]
OK, this (meant to be constructive) issue has become a dumping ground for complaints, so I'm taking the liberty of locking it. If you have questions or concerns, you are welcome to air them on Matrix.

To summarize:

* the original legacy _implementation_ was broken in Neovim and impossible to maintain, so it had to be removed (even if you personally didn't use any modern enough features of Neovim to notice);
* we do want to bring back the _functionality_ of printing styled buffer contents in a more flexible and maintainable form (i.e., as a bundled Lua plugin; if that requires hooks in the C sources, those can be added of course);
* but this requires somebody who actually cares about this functionality to drive this and open a PR; for the project itself this is on the very bottom of the (already over-long) priority list as none of the maintainers used it or see any value for themselves.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/27097 adds treesitter (and more) support to `:TOhtml`, which opens the door for a `:Hardcopy`.
